Sources/Classes/RSContext.m

```diff
@@ -34,34 +42,40 @@ - (instancetype)init
         _timezone = [[NSTimeZone localTimeZone] name];
         _externalIds = nil;
         
-        NSString *traitsJson = [preferenceManager getTraits];
-        if (traitsJson == nil) {
-            // no persisted traits, create new and persist
-            [self createAndPersistTraits];
-        } else {
-            NSError *error;
-            NSDictionary *traitsDict = [NSJSONSerialization JSONObjectWithData:[traitsJson dataUsingEncoding:NSUTF8StringEncoding] options:NSJSONReadingMutableContainers error:&error];
-            if (error == nil) {
-                _traits = [traitsDict mutableCopy];
-            } else {
-                // persisted traits persing error. initiate blank
+        dispatch_sync(queue, ^{
+            NSString *traitsJson = [preferenceManager getTraits];
+            if (traitsJson == nil) {
+                // no persisted traits, create new and persist
                 [self createAndPersistTraits];
+            } else {
+                NSError *error;
+                NSDictionary *traitsDict = [NSJSONSerialization JSONObjectWithData:[traitsJson dataUsingEncoding:NSUTF8StringEncoding] options:NSJSONReadingMutableContainers error:&error];
+                if (error == nil) {
+                    _traits = [traitsDict mutableCopy];
+                } else {
+                    // persisted traits persing error. initiate blank
+                    [self createAndPersistTraits];
+                }
             }
-        }
         
-        // get saved external Ids from prefs
-        NSString *externalIdJson = [preferenceManager getExternalIds];
-        if (externalIdJson != nil) {
-            NSError *error;
-            NSDictionary *externalIdDict = [NSJSONSerialization JSONObjectWithData:[externalIdJson dataUsingEncoding:NSUTF8StringEncoding] options:NSJSONReadingMutableContainers error:&error];
-            if (error == nil) {
-                _externalIds = [externalIdDict mutableCopy];
+            // get saved external Ids from prefs
+            NSString *externalIdJson = [preferenceManager getExternalIds];
+            if (externalIdJson != nil) {
+                NSError *error;
+                NSDictionary *externalIdDict = [NSJSONSerialization JSONObjectWithData:[externalIdJson dataUsingEncoding:NSUTF8StringEncoding] options:NSJSONReadingMutableContainers error:&error];
+                if (error == nil) {
+                    _externalIds = [externalIdDict mutableCopy];
```

Comment:
`_externalIds` is an `array`. You are assigning a `dictionary` to it.

Reply:
It is the existing implementation, I haven't made any improvements on it. I have just synchronised the block. Should we need to make any changes?

Reply:
Yes please.